README.md
# What is Mimic?
Mimic is a testing tool that lets you set create a fake stand-in for an external web service to be used when writing integration/end-to-end tests for applications or libraries that access these services.

## Why not stub?
There are already some good tools, like [FakeWeb](http://fakeweb.rubyforge.org/) which let you stub requests at a low-level which is fine for unit and functional tests but when exercising our code through integration or end-to-end tests we want to exercise as much of the stack as possible.

Mimic aims to make it possible to test your networking code without actually hitting the real services by starting up a real web server and responding to HTTP requests. This lets you test your application against canned responses in an as-close-to-the-real-thing-as-possible way.

Also, because Mimic responds to real HTTP requests, it can be used when testing non-Ruby applications too.

## Examples

Registering to a single request stub:

    Mimic.mimic.get("/some/path").returning("hello world")
    
And the result, using RestClient:
  
    $ RestClient.get("http://www.example.com:11988/some/path") # => 200 | hello world
  
Registering multiple request stubs; note that you can stub the same path with different HTTP methods separately.

    Mimic.mimic do
      get("/some/path").returning("Hello World", 200)
      get("/some/other/path").returning("Redirecting...", 301, {"Location" => "somewhere else"})
      post("/some/path").returning("Created!", 201)
    end
    
You can even use Rack middlewares, e.g. to handle common testing scenarios such as authentication:

    Mimic.mimic do
      use Rack::Auth::Basic do |user, pass|
        user == 'theuser' and pass == 'thepass'
      end
      
      get("/some/path")
    end
    
Finally, because Mimic is built on top of Sinatra for the core request handling, you can create your stubbed requests like you would in any Sinatra app:

    Mimic.mimic do
      get "/some/path" do
        [200, {}, "hello world"]
      end
    end

## Using Mimic with non-Ruby processes

Mimic has a built-in REST API that lets you configure your request stubs over HTTP. This makes it possible to use Mimic from other processes that can perform HTTP requests.

First of all, you'll need to run Mimic as a daemon. You can do this with a simple Ruby script:

    #!/usr/bin/env ruby
    Mimic.daemonize({:port => 11988, :remote_configuration_path => '/api'})
    
Give the script executable permissions and then start it:

    $ your_mimic_script.rb start (or run)
    
The remote configuration path is where the API endpoints will be mounted - this is configurable as you will not be able this path or any paths below it in your stubs, so choose one that doesn't conflict with the paths you need to stub.

The API supports both JSON and Plist payloads, defaulting to JSON. Set the request Content-Type header to application/plist for Plist requests.

For the following Mimic configuration (using the Ruby DSL):

    Mimic.mimic.get("/some/path").returning("hello world")
    
The equivalent stub can be configured using the REST API as follows:

    $ curl -d'{"path":"/some/path", "body":"hello world"}' http://localhost:11988/api/get
    
Likewise, a POST request to the same path could be stubbed like so:

    $ curl -d'{"path":"/some/path", "body":"hello world"}' http://localhost:11988/api/post

The end-point of the API is the HTTP verb you are stubbing, the path, response body, code and headers are specified in the POST data (a hash in JSON or Plist format). See the HTTP API Cucumber features for more examples.

## Contributors

* [James Fairbairn](http://github.com/jfairbairn)
* [Marcello Barnaba](https://github.com/vjt)

## License

As usual, the code is released under the MIT license which is included in the repository.



CHANGES
## 0.4.3

* Added request logging support.

## 0.4.2

* Remove unnecessary signal traps.

## 0.4.1

* Fixed bug where stubs wouldn't actually be cleared correctly.

## 0.4.0

* Mimic can now be run in it's own process and configured externally using a REST API.
* Mimic can be run using the Daemons gem safely (pass :fork => false to disable built-in forking).
* All existing stubs can be cleared.

## 0.3.0

* All verb methods (get, post etc.) can take blocks

## 0.2.0

* Added support for using Rack middleware
* Removed host file modification feature
* Refactor code top build on top of Sinatra::Base

## 0.1.0

* Initial release


wrappers/Objective-C/LRMimic.m
//
//  LRMimic.m
//  MimicSpecDemo
//
//  Created by Luke Redpath on 13/01/2011.
//  Copyright 2011 LJR Software Limited. All rights reserved.
//

#import "LRMimic.h"
#import "LRResty.h"

@interface LRMimic ()
+ (id)sharedInstance;
- (NSData *)payload;
@end

#define kMimicDefaultURL @"http://localhost:11988/api"

@implementation LRMimic

static NSString *mimicURL = nil;
static BOOL automaticallyClearStubs = NO;

+ (void)setURL:(NSString *)url;
{
  [mimicURL autorelease];
  mimicURL = [url copy];
}

+ (void)setAutomaticallyClearsStubs:(BOOL)shouldClear
{
  automaticallyClearStubs = shouldClear;
}

+ (void)reset;
{
  [[self sharedInstance] reset];
}

#pragma mark -
#pragma mark Static API

+ (void)configure:(LRMimicConfigurationBlock)configurationBlock
{
  configurationBlock([self sharedInstance]);
}

+ (void)stubAndCall:(LRMimicCallback)callback;
{
  [[self sharedInstance] prepareStubs:callback clearRemote:automaticallyClearStubs];
}

#pragma mark -

- (id)initWithMimicURL:(NSString *)mimicURL;
{
  if (self = [super init]) {
    URL = [mimicURL copy];
    stubs = [[NSMutableArray alloc] init];
    client = [LRResty newClient];

    [client attachRequestModifier:^(LRRestyRequest *req) {
      [req addHeader:@"Content-Type" value:@"application/plist"];
    }];
  }
  return self;
}

- (void)dealloc
{
  [client release];
  [mimicURL release];
  [stubs release];
  [super dealloc];
}

- (void)prepareStubs:(LRMimicCallback)callback;
{
  [client post:[URL stringByAppendingPathComponent:@"multi"]
       payload:[self payload]
     withBlock:^(LRRestyResponse *response) {
       
    if (response.status == 201) {  
      callback(YES);
    } else {
      callback(NO);
    }
  }];
}
  
- (void)prepareStubs:(LRMimicCallback)callback clearRemote:(BOOL)clearRemote;
{
  if (clearRemote) {
    [self clearRemoteStubs:^(BOOL success) {
      if (success) {
        [self prepareStubs:callback];
      } else {
        callback(NO);
      }
    }];
  } else {
    [self prepareStubs:callback];
  }
}

- (void)clearRemoteStubs:(LRMimicCallback)callback;
{
  [client post:[URL stringByAppendingPathComponent:@"clear"]
       payload:nil
     withBlock:^(LRRestyResponse *response) {
       
    if (response.status == 200) {
      callback(YES);
    } else {
      callback(NO);
    }
  }];
}

- (void)reset;
{
  [stubs removeAllObjects];
}

#pragma mark -
#pragma mark Private methods

- (NSData *)payload;
{
  NSMutableArray *stubPayloads = [NSMutableArray arrayWithCapacity:stubs.count];
  for (LRMimicRequestStub *stub in stubs) {
    [stubPayloads addObject:[stub toDictionary]];
  }
  NSLog(@"STUBBING %@", stubPayloads);
  NSDictionary *payload = [NSDictionary dictionaryWithObject:stubPayloads forKey:@"stubs"];
  return [NSPropertyListSerialization dataFromPropertyList:(id)payload
                                                    format:NSPropertyListXMLFormat_v1_0 
                                          errorDescription:nil];
}

+ (id)sharedInstance;
{
  static id sharedInstance = nil;
  
  if (sharedInstance == nil) {
    if (mimicURL == nil) {
      mimicURL = kMimicDefaultURL;
    }
    sharedInstance = [[self alloc] initWithMimicURL:mimicURL];
  }
  
  return sharedInstance;
}

@end

#pragma mark -

@implementation LRMimic (StubMethods)

- (LRMimicRequestStub *)get:(NSString *)path;
{
  return [self stub:path method:@"GET"];
}

- (LRMimicRequestStub *)post:(NSString *)path;
{
  return [self stub:path method:@"POST"];
}

- (LRMimicRequestStub *)put:(NSString *)path;
{
  return [self stub:path method:@"PUT"];
}

- (LRMimicRequestStub *)delete:(NSString *)path;
{
  return [self stub:path method:@"DELETE"];
}

- (LRMimicRequestStub *)head:(NSString *)path;
{
  return [self stub:path method:@"HEAD"];
}

- (LRMimicRequestStub *)stub:(NSString *)path method:(NSString *)httpMethod;
{
  LRMimicRequestStub *stub = [LRMimicRequestStub stub:path method:httpMethod];
  [self addRequestStub:stub];
  return stub;
}

- (void)addRequestStub:(LRMimicRequestStub *)stub;
{
  [stubs addObject:stub];
}

@end


#pragma mark -

@implementation LRMimicRequestStub

+ (id)stub:(NSString *)path;
{
  return [self stub:path method:@"GET"];
}

+ (id)stub:(NSString *)path method:(NSString *)method
{
  return [[[self alloc] initWithPath:path method:method] autorelease];
}

- (id)initWithPath:(NSString *)aPath method:(NSString *)HTTPMethod;
{
  if (self = [super init]) {
    method = [HTTPMethod copy];
    path = [aPath copy];
    body = [[NSString alloc] initWithString:@""];
    code = 200;
    shouldEchoRequest = NO;
  }
  return self;
}

- (void)dealloc
{
  [queryParameters release];
  [headers release];
  [method release];
  [path release];
  [body release];
  [super dealloc];
}

- (void)willReturnResponse:(NSString *)responseBody withStatus:(NSInteger)statusCode 
                   headers:(NSDictionary *)theHeaders queryParameters:(NSDictionary *)params;
{
  body = [responseBody copy];
  code = statusCode;
  headers = [theHeaders copy];
  queryParameters = [params copy];
}

- (void)willReturnResponse:(NSString *)responseBody withStatus:(NSInteger)statusCode headers:(NSDictionary *)theHeaders;
{
  [self willReturnResponse:responseBody withStatus:statusCode headers:theHeaders queryParameters:nil];
}

- (void)willReturnResponse:(NSString *)responseBody withStatus:(NSInteger)statusCode;
{
  [self willReturnResponse:responseBody withStatus:statusCode headers:nil];
}

- (void)andEchoRequest;
{
  shouldEchoRequest = YES;
}

- (NSDictionary *)toDictionary;
{
  NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
  [dictionary setObject:method forKey:@"method"];
  [dictionary setObject:path forKey:@"path"];
  [dictionary setObject:body forKey:@"body"];
  [dictionary setObject:[NSNumber numberWithInteger:code] forKey:@"code"];
  if (shouldEchoRequest) {
    [dictionary setObject:@"plist" forKey:@"echo"];
  }
  if (headers) {
    [dictionary setObject:headers forKey:@"headers"];
  }
  if (queryParameters) {
    [dictionary setObject:queryParameters forKey:@"params"];
  }
  return dictionary;
}

@end

#pragma mark -
#pragma mark Stub builder

@implementation LRMimicRequestStubBuilder

@synthesize path, method, code, body, headers, queryParameters, echoRequest;

+ (id)builder;
{
  return [[[self alloc] init] autorelease];
}

- (id)init
{
  if (self = [super init]) {
    self.path = @"/";
    self.method = @"GET";
    self.code = 200;
    self.body = @"";
    self.headers = [NSDictionary dictionary];
    self.queryParameters = [NSDictionary dictionary];
    self.echoRequest = NO;
  }
  return self;
}

- (void)dealloc
{
  [queryParameters release];
  [path release];
  [method release];
  [body release];
  [headers release];
  [super dealloc];
}

- (LRMimicRequestStub *)buildStub;
{
  LRMimicRequestStub *stub = [LRMimicRequestStub stub:self.path method:self.method];
  [stub willReturnResponse:self.body withStatus:self.code headers:self.headers queryParameters:self.queryParameters];
  if (self.echoRequest) {
    [stub andEchoRequest];
  }
  return stub;
}

@end

